### PR TITLE
util: add WaitGroupWrapper

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Weizhen Wang <wangweizhen@pingcap.com>

### What problem does this PR solve?

Problem Summary:

add WaitGroupWrapper

### Check List

Tests 

- [x] No code

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by hawkingrei** [comment]
be relation with #29517

**Review by YuJuncen — APPROVED** [review]

File: util/wait_group_wrapper.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+type WaitGroupWrapper struct {
+	sync.WaitGroup
+}
+
+func (w *WaitGroupWrapper) Wrap(cb func()) {
+	w.Add(1)
+	go func() {
+		cb()
+		w.Done()
```
Comment (kennytm) [inline util/wait_group_wrapper.go:29]: if `cb()` panics then `w.Done()` will never be called and `w.Wait()` will be stuck forever.

**Comment by YuJuncen** [comment]
Just a propose: we can choose the same name `Go` as the `errgroup` did.

**Comment by kennytm** [comment]
perhaps just make that a free function 🤷 

```go
func Go(wg *sync.WaitGroup, cb func()) {
    wg.Add(1)
    go func() {
        defer wg.Done()
        cb()
    }()
}
```

File: util/wait_group_wrapper.go (hunk 1 above)
Comment (hawkingrei) [reply util/wait_group_wrapper.go:29]: I think we can add ```WrapWithPanic```, if panic , it will run ```recover``` to ``` w.Done()```. Just as we do now.

**Comment by hawkingrei** [comment]
> perhaps just make that a free function 🤷
> 
> ```go
> func Go(wg *sync.WaitGroup, cb func()) {
>     wg.Add(1)
>     go func() {
>         defer wg.Done()
>         cb()
>     }()
> }
> ```



> perhaps just make that a free function 🤷
> 
> ```go
> func Go(wg *sync.WaitGroup, cb func()) {
>     wg.Add(1)
>     go func() {
>         defer wg.Done()
>         cb()
>     }()
> }
> ```

I agree, and we can add recovery func in ther ```defer```

```
func Wrap(wg *sync.WaitGroup, cb func(), recovery func()) {
    wg.Add(1)
    go func() {
        defer func() {
		recovery()
		d.wg.Done()
	}()
        cb()
    }()
}
```

and I think ```Wrap``` is better than ```go``` for avoiding confusion it is ```errgroup```.

File: util/wait_group_wrapper.go (hunk 1 above)
Comment (hawkingrei) [reply util/wait_group_wrapper.go:33]: have added ```RunWithRecover```

File: util/wait_group_wrapper.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+// it will dump current goroutine stack into log if catch any recover result.
+//   exec:      execute logic function.
+//   recoverFn: handler will be called after recover and before dump stack, passing `nil` means noop.
+func (w *WaitGroupWrapper) RunWithRecover(exec func(), recoverFn func(r interface{})) {
+	w.Add(1)
+	go func() {
+		defer func() {
+			r := recover()
+			if recoverFn != nil {
+				recoverFn(r)
```
Comment (YuJuncen) [inline util/wait_group_wrapper.go:51]: Seems the recoverFn would always run even `exec` didn't panic, is it OK? 

See https://play.golang.org/p/s5gxgKCwFBN

File: util/wait_group_wrapper.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+// Run run a function in a goroutine, add 1 to WaitGroup
+// and call done when function return. Please not use panic
+// in the cb function.
+func (w *WaitGroupWrapper) Run(cb func()) {
+	w.Add(1)
+	go func() {
+		cb()
+		w.Done()
```
Comment (YuJuncen) [inline util/wait_group_wrapper.go:36]: BTW, I guess kenny means call `w.Done()` by `defer w.Done()`(which would be executed even `cb` panics).

File: util/wait_group_wrapper.go (hunk 2 above)
Comment (hawkingrei) [reply util/wait_group_wrapper.go:48]: Fix

File: util/wait_group_wrapper.go (hunk 3 above)
Comment (hawkingrei) [reply util/wait_group_wrapper.go]: OK, I have fixed.

File: util/wait_group_wrapper.go (hunk 2 above)
Comment (hawkingrei) [reply util/wait_group_wrapper.go:48]: I have fixed and push the github. but I don't know why the old version of the code is still displayed.

File: util/wait_group_wrapper.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+// WaitGroupWrapper is a wrapper for sync.WaitGroup
+type WaitGroupWrapper struct {
+	sync.WaitGroup
+}
+
+// Run run a function in a goroutine, add 1 to WaitGroup
+// and call done when function return. Please not use panic
+// in the cb function.
+func (w *WaitGroupWrapper) Run(cb func()) {
```
Comment (dragonly) [inline util/wait_group_wrapper.go:29]: suggest use `cb` or `exec` in both function (consistent argument name).

**Review by dragonly — CHANGES_REQUESTED** [review]

File: util/wait_group_wrapper.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+		defer w.Done()
+		cb()
+	}()
+}
+
+// RunWithRecover wraps goroutine startup call with force recovery, add 1 to WaitGroup
+// and call done when function return.
+// it will dump current goroutine stack into log if catch any recover result.
+//   exec:      execute logic function.
+//   recoverFn: handler will be called after recover and before dump stack, passing `nil` means noop.
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply util/wait_group_wrapper.go:41]: It is copied from ```util.WithRecovery```.

File: util/wait_group_wrapper.go (hunk 4 above)
Comment (hawkingrei) [reply util/wait_group_wrapper.go]: Agree with you. fixed

**Review by ichn-hu — COMMENTED** [review]
LGTM, but how can we let all engineers know about this?

**Review by ichn-hu — APPROVED** [review]

**Pull request merged**